File: src/main/java/dev/algos/snatch/interview_problems/dp/MinimumCostToMergeStones.java
package dev.algos.snatch.interview_problems.dp;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * There are N piles of stones arranged in a row.  The i-th pile has stones[i] stones.
 * <p>
 * A move consists of merging exactly K consecutive piles into one pile, and the cost of this move is equal to the total number of stones in these K piles.
 * <p>
 * Find the minimum cost to merge all piles of stones into one pile.  If it is impossible, return -1.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * Input: stones = [3,2,4,1], K = 2
 * Output: 20
 * Explanation:
 * We start with [3, 2, 4, 1].
 * We merge [3, 2] for a cost of 5, and we are left with [5, 4, 1].
 * We merge [4, 1] for a cost of 5, and we are left with [5, 5].
 * We merge [5, 5] for a cost of 10, and we are left with [10].
 * The total cost was 20, and this is the minimum possible.
 */
public class MinimumCostToMergeStones {

    /**
     * Naive solution
     * TLE suka
     */
    public int mergeStones(int[] stones, int k) {
        Map<List<Integer>, Integer> memo = new HashMap<>();
        List<Integer> list = new LinkedList<>();
        for (int stone : stones) {
            list.add(stone);
        }
        return mergeRec(list, k, memo);
    }

    int mergeRec(List<Integer> list, int k, Map<List<Integer>, Integer> memo) {
        if (list.size() == 1) return 0;
        if (list.size() < k) return -1;

        if (memo.containsKey(list)) return memo.get(list);

        int cost = Integer.MAX_VALUE;
        for (int i = 0; i <= list.size() - k; i++) {
            List<Integer> sub = new LinkedList<>();
            int subSum = 0;
            for (int j = i; j < i + k; j++) {
                sub.add(list.get(i));
                subSum += list.get(i);
                list.remove(i);
            }
            list.add(i, subSum);
            int subCost = mergeRec(list, k, memo);
            if (subCost == -1) return -1;
            cost = Math.min(cost, subSum + subCost);
            list.remove(i);
            list.addAll(i, sub);
        }
        memo.put(list, cost);
        return cost;
    }
}

File: src/main/java/dev/algos/snatch/interview_problems/miscellaneous/MinimumCostToMergeStones.kt
package dev.algos.snatch.interview_problems.miscellaneous

import kotlin.math.min

class MinimumCostToMergeStones {

    /*
    We keep merging K piles of stones until there is only one pile.
    For the last step, stones[i .. j] are divided into K piles, and we merge them into one pile,
    which costs sum(nums[i .. j]) + cost to make stones[i .. j] form K piles.
    The problem get the minimum cost to make stones[i .. j] form 1 pile equals to
	the minimum cost to make stones[i .. j] form K piles + sum(nums[i .. j])

    The subproblem the minimum cost to make stones[i .. j] form K piles equals to
	the minimum cost to make stones[i .. k] form K - 1 piles
	+ the minimum cost to make stones[k + 1 .. j] form 1 pile
	+ sum(nums[i .. j])

    Input: stones = [3,2,4,1], K = 2
    Output: 20
    Explanation:
    We start with [3, 2, 4, 1].

    We merge [3, 2] for a cost of 5, and we are left with [5, 4, 1].
    We merge [4, 1] for a cost of 5, and we are left with [5, 5].
    We merge [5, 5] for a cost of 10, and we are left with [10].
    The total cost was 20, and this is the minimum possible.

    3 2 4 1

     */


    fun mergeStones(stones: IntArray, k: Int): Int {
        val prefix = stones.scan(0, Int::plus)
        val dp = Array(stones.size + 1) { Array(stones.size + 1) { IntArray(k + 1) } }
        val min = helper(1, stones.size, 1, k, stones, prefix, dp)
        return if (min == Int.MAX_VALUE) -1 else min
    }

    // Minimum cost merging piles from i to j into targetPiles pile.
    private fun helper(i: Int, j: Int, piles: Int, k: Int, stones: IntArray, prefix: List<Int>, dp: Array<Array<IntArray>>): Int {
        if (j - i + 1 < piles) return Int.MAX_VALUE
        if (i == j) return if (piles == 1) 0 else Int.MAX_VALUE
        if (dp[i][j][piles] == 0) {
            if (piles == 1) {
                val sub = helper(i, j, k, k, stones, prefix, dp)
                dp[i][j][piles] = if (sub != Int.MAX_VALUE) prefix[j] - prefix[i - 1] + sub else sub
            } else {
                var min = Int.MAX_VALUE
                for (m in i until j) {
                    val left = helper(i, m, piles - 1, k, stones, prefix, dp)
                    if (left == Int.MAX_VALUE) continue
                    val right = helper(m + 1, j, 1, k, stones, prefix, dp)
                    if (right == Int.MAX_VALUE) continue
                    min = min(left + right, min)
                }
                dp[i][j][piles] = min
            }
        }
        return dp[i][j][piles]
    }
}
